permissions/src/test/java/org/odk/collect/permissions/PermissionsProviderTest.kt

```diff
@@ -23,118 +22,299 @@ class PermissionsProviderTest {
 
     private lateinit var permissionsProvider: PermissionsProvider
 
     @Before
     fun setup() {
-        permissionsProvider = PermissionsProvider(permissionsChecker, permissionsApi)
+        permissionsProvider = PermissionsProvider(permissionsChecker, permissionsApi, permissionsDialogCreator, locationAccessibilityChecker)
     }
 
     @Test
-    fun `When camera permission granted should isCameraPermissionGranted() return true`() {
+    fun `When camera permission granted should isCameraPermissionGranted return true`() {
         whenever(permissionsChecker.isPermissionGranted(Manifest.permission.CAMERA)).thenReturn(true)
 
         assertThat(permissionsProvider.isCameraPermissionGranted, `is`(true))
     }
 
     @Test
-    fun `When camera permission not granted should isCameraPermissionGranted() return false`() {
-        whenever(permissionsChecker.isPermissionGranted(Manifest.permission.CAMERA))
-            .thenReturn(false)
+    fun `When camera permission is not granted should isCameraPermissionGranted return false`() {
+        whenever(permissionsChecker.isPermissionGranted(Manifest.permission.CAMERA)).thenReturn(false)
 
         assertThat(permissionsProvider.isCameraPermissionGranted, `is`(false))
     }
 
     @Test
-    fun `When location permissions granted should areLocationPermissionsGranted() return true`() {
-        whenever(
-            permissionsChecker.isPermissionGranted(
-                Manifest.permission.ACCESS_FINE_LOCATION,
-                Manifest.permission.ACCESS_COARSE_LOCATION
-            )
-        ).thenReturn(true)
+    fun `When camera permission is granted should requestCameraPermission call PermissionListener#granted`() {
+        permissionsApi.setGrantedPermission(Manifest.permission.CAMERA)
+
+        permissionsProvider.requestCameraPermission(activity, permissionListener)
+
+        verify(permissionListener).granted()
+        verifyNoMoreInteractions(permissionListener)
+        verifyNoInteractions(permissionsDialogCreator)
+    }
+
+    @Test
+    fun `When camera permission is not granted should requestCameraPermission call PermissionListener#denied`() {
+        permissionsProvider.requestCameraPermission(activity, permissionListener)
+
+        verify(permissionListener).denied()
+        verifyNoMoreInteractions(permissionListener)
+    }
+
+    @Test
+    fun `When camera permission is not granted should requestCameraPermission call PermissionsDialogCreator#showAdditionalExplanation`() {
+        permissionsProvider.requestCameraPermission(activity, permissionListener)
+
+        verify(permissionsDialogCreator).showAdditionalExplanation(
+            activity,
+            R.string.camera_runtime_permission_denied_title,
+            R.string.camera_runtime_permission_denied_desc,
+            R.drawable.ic_photo_camera,
+            permissionListener
+        )
+        verifyNoMoreInteractions(permissionsDialogCreator)
```

Comment:
It might be worth investigating using the Mockito JUnit rule. I think that allows you to use "strict stubs" that will fail tests if non-verified interactions occur (meaning you don't need to add explicit `verifyNoMoreInteractions`/`verifyNoInteractions`). Looking at [SO](https://stackoverflow.com/questions/8003278/is-it-possible-to-do-strict-mocks-with-mockito), it seems we can do that with:

```kotlin
MockitoJUnit.rule().strictness(Strictness.STRICT_STUBS)
```

Reply:
I gave it a quick try but it didn't work well. If you want to play with it please go ahead as you know I won't be able to do that soon.